[clang] Fix the crash when output file is using absolute path

cachemeifyoucan opened the pull request:
Fix the problem that clang -fdepscan cannot handle absolute path as
output by not hard code the output path in the CAS. Instead, let
cc1_main to remap the file path using the information in compiler
invocation.

rdar://86215990

dexonsmith commented on clang/test/CAS/fdepscan-result-cache-output-path.c:
```diff
@@ -0,0 +1,15 @@
+//
+/// Test absolute path output.
+// RUN: rm -rf %t && mkdir %t
+// RUN: %clang -target x86_64-apple-macos11 -fdepscan=daemon -c -x c %s -o %t/out.o
```
Can we use tests `-fdepscan=inline` here to avoid unnecessary extra processes when running tests?

dexonsmith commented on clang/test/CAS/fdepscan-result-cache-output-path.c:
```diff
@@ -0,0 +1,15 @@
+// TODO: Test should be updated to not depend on a working cache at default location, which is out of test/build directory.
+//
+/// Test absolute path output.
+// RUN: rm -rf %t && mkdir %t
+// RUN: %clang -target x86_64-apple-macos11 -fdepscan=inline -c -x c %s -o %t/out.o
+// RUN: ls %t/out.o && rm %t/out.o
+// RUN: %clang -target x86_64-apple-macos11 -fdepscan=inline -c -x c %s -o %t/out.o
```
It's not clear why this is being run twice, but it would be good to check whether there's a cache hit/miss with `-Rresult-cache`. Relatedly, I think @akyrtzi landed support for `-fcas-builtin-path` with `-fdepscan`. That means you can use a fresh CAS in `%t` and get a known miss on the first result.

dexonsmith commented on clang/test/CAS/fdepscan-result-cache-output-path.c:
```diff
@@ -0,0 +1,15 @@
+//
+/// Test relative path output.
+// RUN: cd %t
+// RUN: %clang -target x86_64-apple-macos11 -fdepscan=inline -c -x c %s -o out.o
+// RUN: ls out.o && rm out.o
```
If the input is canonicalized, we should be able to get a cache hit here, since we've already done this with an absolute path.

dexonsmith commented on llvm/include/llvm/CAS/CASOutputBackend.h:
```diff
@@ -11,8 +11,11 @@
 
 #include <llvm/Support/Error.h>
 #include <llvm/Support/OutputBackend.h>
+#include <memory>
```
Is this for `std::shared_ptr<>` below? I suggest just pushing this separately (no need for a review).

dexonsmith commented on clang/tools/driver/cc1_main.cpp:
```diff
@@ -407,8 +446,9 @@ int cc1_main(ArrayRef<const char *> Argv, const char *Argv0, void *MainAddr) {
 
     // Set up the output backend so we can save / cache the result after.
+    auto Mapper = createCASPrefixMapperFromInvocation(Clang->getInvocation());
     CASOutputs = llvm::makeIntrusiveRefCnt<llvm::cas::CASOutputBackend>(
-        *CAS, std::move(CASIDFileOutBackend));
+        *CAS, std::move(CASIDFileOutBackend), Mapper);
```
I think the principled approach would move code similar to this to something before `createResultCacheKey()`. Probably it needs to:
- Modify the CompilerInvocation.
- Save some side state for replaying a cache hit and/or filtering/decanonicalizing during a cache miss.
- Create a new Argv array to send to `createResultCacheKey()` based on the pruned/canonicalized CompilerInvocation.

dexonsmith commented on clang/test/CAS/fdepscan-result-cache-output-path.c:
```diff
@@ -0,0 +1,15 @@
+//
+/// Test absolute path output.
+// RUN: rm -rf %t && mkdir %t
+// RUN: %clang -target x86_64-apple-macos11 -fdepscan=inline -c -x c %s -o %t/out.o
```
I don't feel like the dep-scanner is relevant for this test; is it? I'm wondering if you can use `-cc1`, after calling `llvm-cas --ingest` with some args to create the necessary inputs?

dexonsmith reviewed (changes requested):
This looks like it mostly works, but it's not following a principle I think is critical: higher level computations should canonicalize *inputs* to low-level computations. (This way, as you go deeper down the hierarchy, finer-grained computations get more cache hits.)

In this context, the input to the low-level computation is a CompilerInvocation, but you haven't canonicalized it. Low-level computations are still seeing the original output file locations.

There are a number of options for how to canonicalize these output locations:
- Don't canonicalize (then no cache hits if they change)
- Make them relative
- Make them some "standard name"

cachemeifyoucan commented:
The goal of this patch is not to improve cache hit by canonicalize the output path but simple fix the problem of handling absolute path but I want to do it in a way that going towards to right direction so it can be used to dealing with the path problem.

If you think I should handle that in the same commit, I will get that implemented soon.

dexonsmith commented:
> The goal of this patch is not to improve cache hit by canonicalize the output path but simple fix the problem of handling absolute path but I want to do it in a way that going towards to right direction so it can be used to dealing with the path problem.

I suggest having `CASOutputBackend` record whether it's a relative vs. absolute path. It can override resolveOutputPath() and prepend `relative/` for relative paths and `absolute/` for absolute paths. Then when replaying you put all the files in `outputs/relative/` somewhere relative and the ones under `outputs/absolute/` at the root.

cachemeifyoucan committed:
[clang] Fix the crash when output file is using absolute path

Fix the problem that clang -fdepscan cannot handle absolute path as
output.

rdar://86215990

shahmishal commented:
@swift-ci test

dexonsmith reviewed (approved):
LGTM

dexonsmith commented:
> @swift-ci test

@shahmishal , looks like the tests fail because whatever is running is looking for Swift-related repos. These PRs should be running llvm-project tests instead.
